File: frontend/src/types/job.ts
export interface MatchBreakdown {
  skills_score: number;
  experience_score: number;
  role_relevance_score: number;
  location_score: number;
}

export interface JobMatchScore {
  job_id: string;
  match_score: number;
  breakdown: MatchBreakdown;
  matched_skills: string[];
  missing_skills: string[];
}

export interface JobListItem {
  id: string;
  title: string;
  company: string;
  location: string | null;
  source: string;
  source_url: string | null;
  salary_min: number | null;
  salary_max: number | null;
  experience_level: string | null;
  job_type: string | null;
  remote_policy: string | null;
  scraped_at: string | null;
  created_at: string | null;
  match_score: number | null;
  is_saved: boolean;
}

export interface JobListResponse {
  jobs: JobListItem[];
  total: number;
}

export interface JobDetail {
  id: string;
  source: string;
  source_url: string | null;
  external_id: string | null;
  title: string;
  company: string;
  description: string | null;
  salary_min: number | null;
  salary_max: number | null;
  salary_currency: string;
  location: string | null;
  experience_level: string | null;
  job_type: string | null;
  remote_policy: string | null;
  ats_platform: string | null;
  apply_url: string | null;
  scraped_at: string | null;
  created_at: string | null;
  alternate_sources: Record<string, unknown>[] | null;
  match_score: number | null;
  match_breakdown: MatchBreakdown | null;
  is_saved: boolean;
}

export interface SavedJob {
  application_id: string;
  job_id: string;
  title: string;
  company: string;
  location: string | null;
  source: string;
  saved_at: string | null;
}

export interface SavedJobsResponse {
  jobs: SavedJob[];
  total: number;
}

export interface MatchStatus {
  status: string;
  scored_count: number;
  total_count: number;
  results: JobMatchScore[] | null;
}

export interface DiscoverRequest {
  keywords?: string[] | null;
  location?: string | null;
  sources?: string[] | null;
}

export interface DiscoverResponse {
  job_id: string;
  message: string;
}

export interface DiscoverStatusResponse {
  status: string;
  last_scraped_at: string | null;
}

export interface SourceStatus {
  name: string;
  type: string;
  available: boolean;
  detail: string | null;
}

export interface SourcesStatusResponse {
  sources: SourceStatus[];
}

File: backend/app/models/job.py
from datetime import datetime

from sqlalchemy import Column, Index, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSON
from sqlmodel import Field

from app.models.base import ExperienceLevel, JobSource, RemotePolicy, TimestampModel


class Job(TimestampModel, table=True):
    __tablename__ = "jobs"
    __table_args__ = (
        UniqueConstraint(
            "source",
            "external_id",
            name="uq_jobs_source_external_id",
        ),
        Index("ix_jobs_title", "title"),
        Index("ix_jobs_company", "company"),
        Index("ix_jobs_location", "location"),
    )

    source: JobSource
    source_url: str | None = Field(default=None)
    external_id: str | None = Field(default=None, index=True)
    title: str
    company: str
    description: str | None = Field(default=None)
    salary_min: int | None = Field(default=None)
    salary_max: int | None = Field(default=None)
    salary_currency: str = Field(default="USD")
    location: str | None = Field(default=None)
    experience_level: ExperienceLevel | None = Field(default=None)
    job_type: str | None = Field(default=None)
    remote_policy: RemotePolicy | None = Field(default=None)
    ats_platform: str | None = Field(default=None)
    apply_url: str | None = Field(default=None)
    scraped_at: datetime | None = Field(default=None)
    alternate_sources: list | None = Field(default=None, sa_column=Column(JSON, nullable=True))
